source/MRMesh/MRToolPath.cpp

```diff
@@ -0,0 +1,181 @@
+#if !defined( __EMSCRIPTEN__) && !defined( MRMESH_NO_VOXEL )
+
+#include "MRToolPath.h"
+#include "MRSurfacePath.h"
+#include "MRFixUndercuts.h"
+#include "MROffset.h"
+#include "MRBox.h"
+#include "MRMesh/MRExtractIsolines.h"
+#include <sstream>
+
+namespace MR
+{
+ToolPathResult constantZToolPath( const Mesh& inputMesh, const AffineXf3f& xf, const ToolPathParams& params )
+{
+    const Vector3f normal = Vector3f::plusZ();
+
+    Mesh meshCopy( inputMesh );
+    meshCopy.transform( xf );
+    FixUndercuts::fixUndercuts( meshCopy, normal, params.voxelSize );
+
+    OffsetParameters offsetParams;
+    offsetParams.voxelSize = params.voxelSize;
+
+    auto resMesh = offsetMesh( meshCopy, params.millRadius, offsetParams );
+    if ( !resMesh.has_value() )
+        return {};
+
+    ToolPathResult  res{ .modifiedMesh = *resMesh };
+    const auto& mesh = res.modifiedMesh;
+
+    const auto box = mesh.getBoundingBox();
+    const float safeZ = box.max.z + params.millRadius;
+
+    const auto plane = MR::Plane3f::fromDirAndPt( normal, box.max );
+    const int steps = int( std::floor( ( plane.d - box.min.z ) / params.sectionStep ) );
+
+    Contour3f toolPath{ { 0, 0, safeZ } };
+    res.commands.push_back( { 0, 0 } );
+
+    MeshEdgePoint prevEdgePoint;
+
+    const float critTransitionLengthSq = params.critTransitionLength * params.critTransitionLength;
+    for ( int step = 0; step < steps; ++step )
+    {        
+        for ( const auto& section : extractPlaneSections( mesh, Plane3f{ plane.n, plane.d - params.sectionStep * step } ) )
+        {
+            Polyline3 polyline;
+            polyline.addFromSurfacePath( mesh, section );
+            const auto contours = polyline.contours().front();
+            
+            auto nearestPointIt = section.begin();
+
+            float minDistSq = FLT_MAX;
+
+            if ( prevEdgePoint.e.valid() )
+            {
+                for ( auto it = section.begin(); it < section.end(); ++it )
+                {
+                    float distSq = ( mesh.edgePoint( *it ) - mesh.edgePoint( prevEdgePoint ) ).lengthSq();
+                    if ( distSq < minDistSq )
+                    {
+                        minDistSq = distSq;
+                        nearestPointIt = it;
+                    }
+                }
+            }
+            auto nextEdgePointIt = nearestPointIt;
+            const float sectionStepSq = params.sectionStep * params.sectionStep;
+            do
+            {
+                std::next( nextEdgePointIt ) != section.end() ? ++nextEdgePointIt : nextEdgePointIt = section.begin();
+            } 
+            while ( nextEdgePointIt != nearestPointIt && ( mesh.edgePoint( *nextEdgePointIt ) - mesh.edgePoint( *nearestPointIt ) ).lengthSq() < sectionStepSq );
```

Comment:
do we need to calculate `mesh.edgePoint( *nearestPointIt )` each iteration? mb do it once out of loop?